internal/provider/kubernetes/storage.go

```diff
@@ -643,10 +652,53 @@ func (v *filesystemSource) getPersistentVolumeClaim(
 			"getting kubernetes PersistentVolumeClaim: %w", err,
 		)
 	}
-	// TODO(storage): it might be good to get the source of truth of these
-	// values from the Pod, but might not be possible.
+
+	// We attempt to find the mount path in the charm container by going through
+	// the pod -> container -> volume mount.
+	// If any of the lookup fails, stop the search and return an error.
+	pod, err := client.CoreV1().Pods(v.client.namespace).Get(ctx, instanceID, v1.GetOptions{})
+	if k8serrors.IsNotFound(err) {
+		return jujustorage.FilesystemAttachmentInfo{}, errors.Errorf(
+			"kubernetes Pod %q not found",
+			instanceID,
+		).Add(coreerrors.NotFound)
+	} else if err != nil {
+		return jujustorage.FilesystemAttachmentInfo{}, errors.Errorf(
+			"getting kubernetes Pod %q: %w", instanceID, err,
+		)
+	}
+
+	containerIdx := slices.IndexFunc(pod.Spec.Containers, func(container core.Container) bool {
+		return container.Name == "charm"
+	})
+	if containerIdx == -1 {
+		return jujustorage.FilesystemAttachmentInfo{}, errors.New(
+			"charm container fot found").Add(coreerrors.NotFound)
+	}
+
+	charmContainer := pod.Spec.Containers[containerIdx]
+	volIdx := slices.IndexFunc(pod.Spec.Volumes, func(volume core.Volume) bool {
```

Comment:
I was worried about this and that it wouldn't work for tmpfs filesystems but I realised that this will be fine because this provider doesn't deal with these aspects.